Disable trailing comma by default

## Introduction and background

When I ran `rustfmt` yesterday on some code I was writing, I was once again reminded by the fact that our current defaults are the following:

```
             where_trailing_comma <boolean> Default: false
                                  Put a trailing comma on where clauses
            struct_trailing_comma [Always|Never|Vertical] Default: Vertical
                                  If there is a trailing comma on structs
        struct_lit_trailing_comma [Always|Never|Vertical] Default: Vertical
                                  If there is a trailing comma on literal structs
              enum_trailing_comma <boolean> Default: true
                                  Put a trailing comma on enum declarations
       match_block_trailing_comma <boolean> Default: false
                                  Put a trailing comma after a block based match arm (non-block arms are not affected)
    match_wildcard_trailing_comma <boolean> Default: true
                                  Put a trailing comma after a wildcard arm
```

I know that some of this is ongoing discussion in https://github.com/rust-lang-nursery/fmt-rfcs/issues/34, since it specifically mentions the trailing comma in some situations:

> The last match arm should have a comma if the body is not a block:

I want to take the chance before #34 is merged to broaden that discussion a bit more, and try to reach consensus on trailing commas in general. (I know that they are sometimes _needed_; I consider that a bug and this discussion is more related to the areas where it is syntactically optional, but a common de-facto standard in the Rust community and/or enforced by `rustfmt`.)

I don't claim to be unbiased; I have a specific opinion that I am voicing in this RFC. However, what I do claim is to represent the different standpoints in a factual way. If you feel that I am missing or misrepresenting some argument for either position, please write a comment with the missing part and I'll try to adjust the proposal continously.

## Basic presumption

I hope that (almost) everyone can agree that our current position - to sometimes advocate a trailing comma and sometimes not, as shown above - is not optimal. It feels counterintuitive and also like something you'd have to teach people about ("the style guide recommends it here, but not there, for whatever reason" - not saying that there _aren't_ reasons for the inconsistency, just saying that an inconsistent set of rules in an area is much harder to defend and also harder for people to remember.)

## Arguments used to support the addition of trailing commas

### It's commonly used within the rust community

This is a fact. Servo uses it, `rustc` uses it, probably a bunch of other projects as well. The precedent is there, there's not so much to say about that. There are probably a number of people who like this style, otherwise it wouldn't have been started to be used anyway. Moving away from this style would risk discouraging those people from contributing to the community.

#### Counter-argument

- It's not that common outside the Rust community.
- The current Rust code in existance will hopefully eventually be a very small minority of the cumulative amount of rust code in the years to come. In other words, the vast majority of the rust code eventually written does not yet exist.
- Once `rustfmt` is stable enough, we should be able to reformat larger codebases (like servo and rustc) with it anyway.
- Regarding the "people might like it" part, it's a fact regarding _any_ part of the style guide. The style guide cannot emphasise individual people so much, but should rather focus on the greater good for the community as a whole.

### It makes changing the code easier

This argument is also factually correct; when adding a new entry to an array, you never have to remember adding a trailing comma. Perhaps more importantly so, when rearranging items in an array you don't have to visually scan the array to find which line now has the missing comma (so you can add it and make the code compile again).

#### Counter-argument

While the argument is factually correct, it's pretty much a matter of what you are used to. Good tooling (i.e. instant feedback on the validity of the code, perhaps powered by [rls](https://github.com/jonathandturner/rls) or simiilar) can make this be more of a moot point. If the editor you are using _instantly_ shows you (or hints at) where the missing comma is, this is less of a point as before. (but of course, not all people can be expected to use a single editor/a RLS-powered editor etc.)

### It produces shorter, less verbose diffs

This argument is based on the simple fact that if the _trailing comma_ is already present on the last line of an {array|enum|match|etc}, that line will not have to be edited if you add a new element to the {array|enum|match}. In other words, the diff produced when changing the code will *only* include the relevant line (the new functionality being added), you won't have to touch an adjacent line to make it compile.

This is a very relevant and factual argument, since it avoids the cognitive load put on the brain when parsing the diff. You can focus on the actual change, and have to think less about the noise.

#### Counter-argument

While this is a a factually correct assertion, @bbatsov (the author of Rubocop, a linter for Ruby), wrote that:

> The defaults will not be changed [to enable trailing commas by default] - using VCS diff limitations to justify a coding convention doesn't make sense

Of course, here is where facts end and opinions start: I personally agree with Boris, that this is too weak an argument to justify the current position. Others might, and will, say the opposite (that this is such a strong argument that the current position should be retained).

### It breaks git blame

This is actually a pretty good argument. It will _appear_ as if the person who added the comma was the author of that line even though his/her contribution was only the single character. Can't really say much about this, this is a reasonable, factual argument. (However, it's again letting the limitations of a particular tool mandate the style choices)

## Arguments used to reject the addition of trailing commas

### Opinionated: It looks like a typo

One of [my early encounters](https://github.com/rust-lang/rust/pull/28376) with the Rust community was actually because of this; I read [some code](https://github.com/rust-lang/rust/blob/5de15be5ec9144f6701c8de606fdf83c6eefefef/src/libstd/ffi/c_str.rs#L75) in the Rust compiler source code which struck me as "looking weird", because of the trailing comma. I have been doing programming for about 20 years, of which 14 years of it has been "for a living". Never during these years have I encountered a community that was advocating trailing commas.

I'm quite certain I'm not alone with this experience (which the SO links further down in this RFC indicates; people tend to get confused when they read code written like this). In my eyes, the current position definitely is not "mainstream" enough for a language with high ambitions to become a broad competitor to languages like C, C++ and perhaps Java, C# and others as well. If that's really what we're aiming for, we should try to aim for a style that "feels natural" to most people reading code _without having seen what the style guide advocates_.

Our language is so great because of the other features it provides (immmutability, the whole borrowing concept, proper concurrency etc) that I think _that_ is what should be sticking out when people read our code: our superiority in these areas. Not that we write code that looks like an array element is missing from the end; that someone has forgotten to finish his sentence. Again, extremely opinionated, but it makes people question the code: is this really what the author intended to write? Is there a bug here, did I just discover an unpleasant surprise?

### It is contrary to the recommended style guide/linter defaults for other programming languages

A non-comprehensive on some of the findings I could gather:

* [rubocop](https://github.com/bbatsov/ruby-style-guide#no-trailing-array-commas), a linting tool for Ruby, produces warnings for this syntax with its default config. The issue is discussed more in https://github.com/bbatsov/rubocop/issues/713, https://github.com/bbatsov/rubocop/pull/735 and at length in https://github.com/bbatsov/ruby-style-guide/issues/273. Other style guides like the [Thoughtbot one](https://github.com/thoughtbot/guides/tree/master/style/ruby) advocates trailing comma, but I'd say the former one is the more popular one (and definitely a more comprehensive guide with much more details about the recommendations).
* [eslint](http://eslint.org/), a linting tool for Javascript, has it [disabled by default](http://eslint.org/docs/rules/comma-dangle) but with options for enforcing it for various scenarios (multi-line only, always, etc).
* [tslint](https://palantir.github.io/tslint/rules/trailing-comma/), a linter for Typescript, seems to take the approach to let you configure it either way, but don't seem to advocate a particular default. (tslint is not a _reformatter_ in that sense, so it can take a more relaxed stance on the matter)
* The [PHP manual](http://php.net/manual/en/function.array.php) says that "[h]aving a trailing comma after the last defined array entry, while unusual, is a valid syntax."
* It has been said to be common in Python, but I couldn't find a style guide to back this claim.

Other languages/style guides don't typically _disallow_ it, but it's a common source for people to scratch their heads and wonder "why is this valid C#/Python/etc code?". Take a look at [this](http://stackoverflow.com/questions/11597901/why-does-python-allow-a-trailing-comma-in-list) and [this](https://docs.python.org/2/faq/design.html#why-does-python-allow-commas-at-the-end-of-lists-and-tuples) link (Python), [this](http://stackoverflow.com/questions/8253413/why-c-sharp-4-0-tolerates-trailing-comma-in-anonymous-objects-initialization-cod) link (C#), [this](http://stackoverflow.com/questions/11621917/arrays-with-trailing-commas-inside-an-array-initializer-in-java) link (Java) etc. It is definitely not a universally accepted style, recommended by a large majority of the style guides for programming languages in general or anything like that.

To put it bluntly: if we want to keep making people confused, we should keep the current recommendation as it is.

### It is invalid code in some languages/environments:

* JSON: Valid code

  ```json
  {
    "foo": "bar",
    "baz": "zot"
  }
  ```
  
  Invalid code, parsers will reject this:
  
  ```json
  {
    "foo": "bar",
    "baz": "zot",
  }
  ```

### Opinionated: It's not really intended for humans writing code

The fact that virtually _all_ programming languages (except from JSON, which isn't really a "programming language") allows it seems to be generally motivated by the fact that it's intended for code generation, where it's obviously a whole lot easier for the code generating code to not have to always check "is this the last element of the array being printed to stdout" and so forth. _C# 2005 Programmer's Reference_ ([p. 208](https://books.google.fi/books?id=KwkFRPzuHCAC&pg=PA208&lpg=PA208&dq=trailing+comma+c%23&source=bl&ots=KFmxoTlf48&sig=yVcaXM7P-UlK8aPxal-uuuaHQE0&hl=en&sa=X&ved=0ahUKEwiP0LLsyM7QAhUHOxQKHT7eBgsQ6AEIVDAJ#v=onepage&q=trailing%20comma%20c%23&f=false)) seems to imply this, and [this SO thread](http://stackoverflow.com/questions/2311864/history-of-trailing-comma-in-programming-language-grammars) gives the same impression. It was initially allowed in C back from the days of K&R, and many other languages have included support for it since.

One could argue that if this "feature" is designed for code generators, let the code generators use it and let humans produce code that is more aesthetically pleasing and straightforward.

## Conclusion and proposed adjustment

I find the communities for various programming languages a bit scattered on the subject. Some people prefer this style very strongly, others (like me) dislike it equally strongly and would never ever want to run a tool to reformat our code with the default settings as they are right now. Others are just confused by reading code formatted like this and start to ask questions about it. That, to me, indicates that our current defaults are not so intuitive and hence I suggest that the following settings should be the new default:

```
             where_trailing_comma <boolean> Default: false
                                  Put a trailing comma on where clauses
            struct_trailing_comma [Always|Never|Vertical] Default: Never
                                  If there is a trailing comma on structs
        struct_lit_trailing_comma [Always|Never|Vertical] Default: Never
                                  If there is a trailing comma on literal structs
              enum_trailing_comma <boolean> Default: false
                                  Put a trailing comma on enum declarations
       match_block_trailing_comma <boolean> Default: false
                                  Put a trailing comma after a block based match arm (non-block arms are not affected)
    match_wildcard_trailing_comma <boolean> Default: false
                                  Put a trailing comma after a wildcard arm
```

IMHO, this would be very consistent and easy to understand for anyone: experts and new Rust enthusiasts alike.

Now, you may send the expected flames... 😉

> I am very much used to no trailing commas, as I can't have them in C++ AFAIK.

I have heard others say the same, and I think the answer is "it depends". This is fully valid code:

``` c++
enum some_enum {
    Foo,
    Bar,
    Baz,
};

int some_array[] = {
    1,
    2,
    3,
};
```

...whereas this is not (as @ubsan pointed out early in this thread):

``` c++
int main() {
    printf("%d %d %d", 1, 2, 3,); // Gives the error "foo.cpp:14:32: error: expected expression" with my Apple LLVM version 8.0.0 (clang-800.0.42.1)
}
```

> And one final note, about the look of trailing separators: I personally actually dislike the look of the C / C++ / Json style multiline lists, because the last line is always inconsistent with the other ones. I always go for YAML when it comes to loading human-written data, because having to think about whether or not to put a comma after a list entry, even though it's such a minor thing, annoys me every time.

That's fine, and I don't claim that everyone using "non-trailing comma" is happy about that; sometimes they are limited by the languages they are using, but would prefer to not be. Sorry if I made it sound like that.

----

However, my point that there are certainly those of us who like to not add the trailing comma, who feels that the "editing argument" and the "VCS argument" is not compelling enough. We like the "traditional" way, with no extra comma at the end, and feel that the perceived "extra work" it causes isn't really a problem for us.

I think it all boils down to personal preference at the end, just like Linus Torvalds once said it:

> Coding style is very personal, and I won't **force** my views on anybody

(quoting from [linux/CodingStyle](https://github.com/torvalds/linux/blob/master/Documentation/CodingStyle#L7))

--- a/guide/guide.md
+++ b/guide/guide.md
@@ -100,6 +100,8 @@ struct CRepr {
 }
 ```
 
+## [Guiding principles and rationale](principles.md)
+
 ## [Non-formatting conventions](advice.md)
 
 ## [Cargo.toml conventions](cargo.md)

--- a/guide/principles.md
+++ b/guide/principles.md
@@ -1,0 +1,51 @@
+# Guiding principles and rationale
+
+When deciding on style guidelines, the style team tried to be guided by the
+following principles (in rough priority order):
+
+* readability
+    - scan-ability
+    - avoiding misleading formatting
+    - accessibility - readable and editable by users using the the widest
+      variety of hardware, including non-visual accessibility interfaces
+    - readability of code when quoted in rustc error messages
+
+* aesthetics
+    - sense of 'beauty'
+    - consistent with other languages/tools
+
+* specifics
+    - compatibility with version control practices - preserving diffs,
+      merge-friendliness, etc.
+    - preventing right-ward drift
+    - minimising vertical space
+
+* application
+    - ease of manual application
+    - ease of implementation (in Rustfmt, and in other tools/editors/code generators)
+    - internal consistency
+    - simplicity of formatting rules
+
+
+## Overarching guidelines
+
+Prefer block indent over visual indent. E.g.,
+
+```
+// Block indent
+a_function_call(
+    foo,
+    bar,
+);
+
+// Visual indent
+a_function_call(foo,
+                bar);
+```
+
+This makes for smaller diffs (e.g., if `a_function_call` is renamed in the above
+example) and less rightward drift.
+
+Lists should have a trailing comma when followed by a newline, see the block
+indent example above. This choice makes moving code (e.g., by copy and paste)
+easier and makes smaller diffs.

